hoomd/MeshGroupData.cu

```diff
@@ -0,0 +1,251 @@
+                            unsigned int next_flag,
+                            unsigned int& flag,
+                            group_t* d_pidx_group_table,
+                            const unsigned int pidx_group_table_pitch,
+                            unsigned int* d_scratch_g,
+                            unsigned int* d_scratch_idx,
+                            unsigned int* d_offsets,
+                            CachedAllocator& alloc)
+    {
+    // construct scratch table by expanding the group table by particle index
+    unsigned int block_size = 256;
+    unsigned n_blocks = n_groups / block_size + 1;
+
+    unsigned int group_size_half = group_size/2;
+
+    // reset number of groups
+    hipMemsetAsync(d_n_groups, 0, sizeof(unsigned int) * N);
+
+    hipLaunchKernelGGL(HIP_KERNEL_NAME(gpu_count_mesh_kernel<group_size>),
+                       dim3(n_blocks),
+                       dim3(block_size),
+                       0,
+                       0,
+                       n_groups,
+                       d_group_table,
+                       d_rtag,
+                       d_scratch_idx,
+                       d_scratch_g,
+                       d_n_groups,
+                       max_n_groups,
+                       d_condition,
+                       next_flag);
+
+    // read back flag
+    hipMemcpy(&flag, d_condition, sizeof(unsigned int), hipMemcpyDeviceToHost);
+
+    if (!(flag >= next_flag) && n_groups)
+        {
+        // we are good, fill group table
+        // sort groups by particle idx
+        thrust::device_ptr<unsigned int> scratch_idx(d_scratch_idx);
+        thrust::device_ptr<unsigned int> scratch_g(d_scratch_g);
+#ifdef __HIP_PLATFORM_HCC__
+        thrust::sort_by_key(thrust::hip::par(alloc),
+#else
+        thrust::sort_by_key(thrust::cuda::par(alloc),
+#endif
+                            scratch_idx,
+                            scratch_idx + group_size_half * n_groups,
+                            scratch_g);
+
+        // perform a segmented scan of d_scratch_idx
+        thrust::device_ptr<unsigned int> offsets(d_offsets);
+        thrust::constant_iterator<unsigned int> const_it(1);
+#ifdef __HIP_PLATFORM_HCC__
+        thrust::exclusive_scan_by_key(thrust::hip::par(alloc),
+#else
+        thrust::exclusive_scan_by_key(thrust::cuda::par(alloc),
+#endif
+                                      scratch_idx,
+                                      scratch_idx + group_size_half * n_groups,
+                                      const_it,
+                                      offsets);
+
+        // scatter groups to destinations
+        block_size = 256;
+        n_blocks = (group_size * n_groups) / block_size + 1;
+
+        hipLaunchKernelGGL(gpu_mesh_scatter_kernel<group_size>,
+                           dim3(n_blocks),
+                           dim3(block_size),
+                           0,
+                           0,
+                           n_groups * group_size_half,
```

Comment:
I there a way to not use this `group_size_half` it seem to be the reason we can't reuse this kernel.

Reply:
I haven't removed it but I think I found the problem that fixes the issue.